File: pythonexamples/golden_hammer.py
"""
ANTI-PATTERN: Golden Hammer

Using the same solution/tool for every problem regardless of whether it's appropriate.
"If all you have is a hammer, everything looks like a nail."
"""

import re

class RegexFanatic:
    """Someone who learned regex and now uses it for EVERYTHING"""

    def validate_email(self, email):
        # OK: Regex is appropriate here
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return bool(re.match(pattern, email))

    def is_even(self, number):
        # GOLDEN HAMMER: Using regex to check if a number is even!
        # Should just use: number % 2 == 0
        return bool(re.match(r'^-?\d*[02468]$', str(number)))

    def add_numbers(self, a, b):
        # GOLDEN HAMMER: Using regex for basic arithmetic!
        # Just use: a + b
        result = str(a) + str(b)
        return int(re.sub(r'(\d+)(\d+)', r'\1\2', result))  # Makes no sense!

    def get_list_length(self, items):
        # GOLDEN HAMMER: Using regex to count list items
        # Should just use: len(items)
        string_rep = str(items)
        matches = re.findall(r',', string_rep)
        return len(matches) + 1 if matches or string_rep != '[]' else 0

    def reverse_string(self, text):
        # GOLDEN HAMMER: Using regex for string reversal
        # Should just use: text[::-1]
        result = text
        for i in range(len(text)):
            pattern = r'^(.{' + str(i) + r'})(.)'
            result = re.sub(pattern, r'\2\1', result)
        return result  # This doesn't even work correctly!


class InheritanceEverything:
    """Someone who uses inheritance for every code reuse scenario"""

    pass

# GOLDEN HAMMER: Using inheritance when composition would be better
class Animal:
    def breathe(self):
        return "breathing"

# A car is NOT an animal, but we're using inheritance just to reuse the breathe logic
class Car(Animal):
    """GOLDEN HAMMER: Car inheriting from Animal just to reuse code!"""
    def drive(self):
        # Now we have a car that can breathe...
        return f"driving while {self.breathe()}"

# GOLDEN HAMMER: Everything must be a class!
class MathOperations:
    """Using classes when simple functions would suffice"""

    @staticmethod
    def add(a, b):
        return a + b

    @staticmethod
    def subtract(a, b):
        return a - b

    @staticmethod
    def multiply(a, b):
        return a * b

    # Should just be simple functions, not a class!


# GOLDEN HAMMER: Using design patterns everywhere
class SingletonFactory:
    """Combining Singleton and Factory patterns when you just need a simple function"""
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def create_string(self, text):
        """This is just creating a string... why is it a singleton factory?"""
        return str(text)

    def create_number(self, num):
        """Just returns a number..."""
        return int(num)


# GOLDEN HAMMER: Using a database for everything
class DatabaseForEverything:
    """Someone who puts everything in a database, even temporary data"""

    def __init__(self, db_connection):
        self.db = db_connection

    def calculate_sum(self, a, b):
        # GOLDEN HAMMER: Storing calculation in database instead of just returning it
        result = a + b
        self.db.execute("INSERT INTO calculations (operation, result) VALUES ('add', ?)", (result,))
        return result

    def temporary_variable(self, value):
        # GOLDEN HAMMER: Storing temporary variable in database!
        self.db.execute("INSERT INTO temp_vars (value) VALUES (?)", (value,))
        row = self.db.execute("SELECT value FROM temp_vars WHERE id = LAST_INSERT_ID()").fetchone()
        self.db.execute("DELETE FROM temp_vars WHERE id = LAST_INSERT_ID()")
        return row[0]


# GOLDEN HAMMER: Using microservices for everything
"""
The developer learned about microservices and now wants to split
a simple application into 50 microservices:

- UserEmailValidationService
- UserFirstNameService
- UserLastNameService
- AddTwoNumbersService
- SubtractTwoNumbersService
- StringToUpperCaseService
- StringToLowerCaseService

Each service has its own database, API, and deployment pipeline
for functionality that should just be simple functions!
"""

File: golangexamples/golden_hammer.go
package main

/*
ANTI-PATTERN: Golden Hammer

Using the same solution/tool for every problem regardless of whether it's appropriate.
"If all you have is a hammer, everything looks like a nail."
*/

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// RegexFanatic - Someone who learned regex and now uses it for EVERYTHING
type RegexFanatic struct{}

// ValidateEmail - OK: Regex is appropriate here
func (rf *RegexFanatic) ValidateEmail(email string) bool {
	pattern := `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
	matched, _ := regexp.MatchString(pattern, email)
	return matched
}

// IsEven - GOLDEN HAMMER: Using regex to check if a number is even!
// Should just use: number % 2 == 0
func (rf *RegexFanatic) IsEven(number int) bool {
	pattern := `^-?\d*[02468]$`
	matched, _ := regexp.MatchString(pattern, strconv.Itoa(number))
	return matched
}

// AddNumbers - GOLDEN HAMMER: Using regex for basic arithmetic!
// Should just use: a + b
func (rf *RegexFanatic) AddNumbers(a, b int) int {
	// This makes no sense and doesn't even work correctly
	combined := fmt.Sprintf("%d%d", a, b)
	re := regexp.MustCompile(`(\d+)`)
	matches := re.FindAllString(combined, -1)
	if len(matches) > 0 {
		result, _ := strconv.Atoi(matches[0])
		return result
	}
	return 0
}

// ReverseString - GOLDEN HAMMER: Using regex for string reversal
// Should just use a simple loop or strings.Builder
func (rf *RegexFanatic) ReverseString(text string) string {
	// This doesn't even work correctly!
	result := text
	for i := 0; i < len(text); i++ {
		pattern := fmt.Sprintf(`^(.{%d})(.)`, i)
		re := regexp.MustCompile(pattern)
		result = re.ReplaceAllString(result, "$2$1")
	}
	return result
}

// ChannelForEverything - Someone who uses channels for every data structure
type ChannelForEverything struct{}

// GOLDEN HAMMER: Using channel as a queue when a slice would be better
func (cfe *ChannelForEverything) CreateQueue() chan int {
	// A simple slice would be more appropriate for most queue use cases
	return make(chan int, 100)
}

// GOLDEN HAMMER: Using channel as a set (doesn't even work well)
func (cfe *ChannelForEverything) CreateSet() chan string {
	// A map[string]bool would be the right choice
	return make(chan string, 100)
}

// GOLDEN HAMMER: Using channel as a single value holder
func (cfe *ChannelForEverything) CreateVariable(value int) chan int {
	// Just use a variable!
	ch := make(chan int, 1)
	ch <- value
	return ch
}

// InterfaceForEverything - Someone who creates interfaces for everything
// even when concrete types would be simpler
type InterfaceForEverything struct{}

// GOLDEN HAMMER: Creating interfaces for simple operations
type Adder interface {
	Add(a, b int) int
}

type Subtractor interface {
	Subtract(a, b int) int
}

type Multiplier interface {
	Multiply(a, b int) int
}

// These should just be functions, not interfaces!
type Calculator struct{}

func (c Calculator) Add(a, b int) int      { return a + b }
func (c Calculator) Subtract(a, b int) int { return a - b }
func (c Calculator) Multiply(a, b int) int { return a * b }

// StructForEverything - Someone who uses structs even for simple data
type StructForEverything struct{}

// GOLDEN HAMMER: Creating a struct to return two values
type TwoInts struct {
	First  int
	Second int
}

// Should just return (int, int)
func (sfe *StructForEverything) GetTwoNumbers() TwoInts {
	return TwoInts{First: 1, Second: 2}
}

// GOLDEN HAMMER: Struct for a simple string pair
type StringPair struct {
	Key   string
	Value string
}

// A map or tuple would be better
func (sfe *StructForEverything) GetConfig() StringPair {
	return StringPair{Key: "timeout", Value: "30"}
}

// GoRoutineForEverything - Someone who uses goroutines for everything
type GoRoutineForEverything struct{}

// GOLDEN HAMMER: Using goroutine for a simple function call
func (gfe *GoRoutineForEverything) PrintMessage(msg string) {
	// No need for a goroutine here, adds unnecessary complexity
	go func() {
		fmt.Println(msg)
	}()
}

// GOLDEN HAMMER: Using goroutine for sequential operations
func (gfe *GoRoutineForEverything) ProcessData(data []int) []int {
	// This makes it slower and more complex!
	resultChan := make(chan []int)
	go func() {
		result := make([]int, len(data))
		for i, v := range data {
			result[i] = v * 2
		}
		resultChan <- result
	}()
	return <-resultChan
}

// ReflectionForEverything - Someone who uses reflection when they shouldn't
type ReflectionForEverything struct{}

// GOLDEN HAMMER: Using reflection for simple type check
func (rfe *ReflectionForEverything) IsString(value interface{}) bool {
	// Should just use type assertion: _, ok := value.(string)
	return fmt.Sprintf("%T", value) == "string"
}

// PointerForEverything - Someone who uses pointers everywhere
type PointerForEverything struct{}

// GOLDEN HAMMER: Using pointers for small primitive types unnecessarily
func (pfe *PointerForEverything) AddNumbers(a *int, b *int) *int {
	// Pointers for ints add complexity without benefit
	result := *a + *b
	return &result
}

func (pfe *PointerForEverything) IsEven(n *int) *bool {
	// Returning pointer to bool is usually unnecessary
	result := *n%2 == 0
	return &result
}

// MapForEverything - Someone who uses maps for every data structure
type MapForEverything struct{}

// GOLDEN HAMMER: Using map as a list
func (mfe *MapForEverything) CreateList() map[int]string {
	// A slice []string would be more appropriate
	return make(map[int]string)
}

// GOLDEN HAMMER: Using map to store two values
func (mfe *MapForEverything) GetUserInfo() map[string]string {
	// A struct would be better
	return map[string]string{
		"name":  "John",
		"email": "john@example.com",
	}
}

// ContextForEverything - Someone who passes context.Context everywhere
// even when it's not needed

// GOLDEN HAMMER: Using context for simple value passing
func ProcessWithUnnecessaryContext(data string) string {
	// Context is overkill when you just need to pass a value
	// Should just pass the value as a parameter
	return strings.ToUpper(data)
}

// FactoryPatternForEverything - Using factories for simple object creation
type FactoryPatternForEverything struct{}

// GOLDEN HAMMER: Factory for a simple struct
type Point struct {
	X, Y int
}

type PointFactory struct{}

func NewPointFactory() *PointFactory {
	return &PointFactory{}
}

func (pf *PointFactory) CreatePoint(x, y int) *Point {
	// Just use: &Point{X: x, Y: y}
	// No factory needed!
	return &Point{X: x, Y: y}
}

// SingletonForEverything - Using singleton pattern unnecessarily
var (
	configInstance     *Config
	loggerInstance     *Logger
	validatorInstance  *Validator
	formatterInstance  *Formatter
	converterInstance  *Converter
	parserInstance     *Parser
	// Everything is a singleton!
)

type Config struct{ value string }
type Logger struct{ name string }
type Validator struct{}
type Formatter struct{}
type Converter struct{}
type Parser struct{}

// GOLDEN HAMMER: Making everything a singleton when it doesn't need to be

// MicroservicesForEverything - Someone who splits everything into microservices
/*
GOLDEN HAMMER: The developer learned about microservices and now wants to split
a simple application into 50 microservices:

- UserEmailValidationService
- UserFirstNameService
- UserLastNameService
- AddTwoNumbersService
- SubtractTwoNumbersService
- StringToUpperCaseService
- StringToLowerCaseService

Each service has its own repository, database, API, and deployment pipeline
for functionality that should just be simple functions!
*/

func main() {
	// Examples of golden hammer in action
	rf := &RegexFanatic{}
	fmt.Println("Is 4 even?", rf.IsEven(4)) // Using regex instead of modulo!

	pfe := &PointerForEverything{}
	a, b := 5, 10
	result := pfe.AddNumbers(&a, &b) // Unnecessary pointer complexity
	fmt.Println("Sum:", *result)

	// The golden hammer principle: when all you have is a hammer,
	// everything looks like a nail. Use the right tool for the job!
}
